In [1]:
import numpy as np
import pandas as pd

In [2]:
tennis=pd.read_csv('play_tennis.csv')

In [3]:
tennis

day   outlook  temp humidity    wind play
0    D1     Sunny   Hot     High    Weak   No
1    D2     Sunny   Hot     High  Strong   No
2    D3  Overcast   Hot     High    Weak  Yes
3    D4      Rain  Mild     High    Weak  Yes
4    D5      Rain  Cool   Normal    Weak  Yes
5    D6      Rain  Cool   Normal  Strong   No
6    D7  Overcast  Cool   Normal  Strong  Yes
7    D8     Sunny  Mild     High    Weak   No
8    D9     Sunny  Cool   Normal    Weak  Yes
9   D10      Rain  Mild   Normal    Weak  Yes
10  D11     Sunny  Mild   Normal  Strong  Yes
11  D12  Overcast  Mild     High  Strong  Yes
12  D13  Overcast   Hot   Normal    Weak  Yes
13  D14      Rain  Mild     High  Strong   No

# Entropy of parent

In [4]:
import math

In [5]:
Ed=-(9/14)*math.log2(9/14)-(5/14)*math.log2(5/14)

Ed

0.9402859586706311

# Entropy for Sunny in the   Outlook Column

In [6]:
E1=-(2/5)*math.log2(2/5)-(3/5)*math.log2(3/5)

E1


0.9709505944546686

# Entropy for Overcast in the outlook column

In [7]:
# As all Overcast has play as Yes hence the entropy is 0  
E2=0


# Entropy for Rain in the Outlook column

In [8]:
E3=-(3/5)*math.log2(3/5)-(2/5)*math.log2(2/5)

E3

0.9709505944546686

# Weighted Entropy of Children

In [9]:


We1=(5/14)*E1+(4/14)*E2+(5/14)*E3

We1

0.6935361388961918



# Information Gain on Outlook Column

In [10]:
IG1=Ed-We1

IG1

0.24674981977443933

## Entropy for hot in temperature column

In [11]:
E11=-(2/4)*math.log2(2/4)-(2/4)*math.log2(2/4)

E11

1.0

## Entropy for mild in teperature column

In [12]:
pd.crosstab(tennis['temp'],tennis['play'])

play  No  Yes
temp         
Cool   1    3
Hot    2    2
Mild   2    4

In [13]:
E12=-(4/6)*math.log2(4/6)-(2/6)*math.log2(2/6)

E12

0.9182958340544896

## Entropy for cool in temperature column

In [14]:
E13=-(3/4)*math.log2(3/4)-(1/4)*math.log2(1/4)

E13

0.8112781244591328

## Weighted Entropy of Temperature column

In [15]:

We2=(4/14)*E11+(6/14)*E12+(4/14)*E13

We2

0.9110633930116763


## IG on Temperature

In [16]:
IG2=Ed-We2

IG2

0.02922256565895487

# Entropy of  Wind Column

In [17]:
pd.crosstab(tennis['wind'],tennis['play'])

play    No  Yes
wind           
Strong   3    3
Weak     2    6

In [18]:
E21=-(3/6)*math.log2(3/6)-(3/6)*math.log2(3/6)

E21

1.0

In [19]:
E22=-(6/8)*math.log2(6/8)-(2/8)*math.log2(2/8)

E22

0.8112781244591328

## Weighted Avg on Wind

In [20]:
We3=(6/14)*E21+(8/14)*E22

We3

0.8921589282623617

## IG on Wind

In [21]:
IG3=Ed-We3

IG3

0.04812703040826949

In [22]:
max(IG1,IG2,IG3)

0.24674981977443933

# Hence the initial Split should be based upon the OutLook Column as it's Information Gain is maximum.

In [23]:
iris=pd.read_csv('Iris.csv')

iris

Id  SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm  \
0      1            5.1           3.5            1.4           0.2   
1      2            4.9           3.0            1.4           0.2   
2      3            4.7           3.2            1.3           0.2   
3      4            4.6           3.1            1.5           0.2   
4      5            5.0           3.6            1.4           0.2   
5      6            5.4           3.9            1.7           0.4   
6      7            4.6           3.4            1.4           0.3   
7      8            5.0           3.4            1.5           0.2   
8      9            4.4           2.9            1.4           0.2   
9     10            4.9           3.1            1.5           0.1   
10    11            5.4           3.7            1.5           0.2   
11    12            4.8           3.4            1.6           0.2   
12    13            4.8           3.0            1.4           0.1   
13    14            4.3           3.0            1.1           0.1   
14    15            5.8           4.0            1.2           0.2   
15    16            5.7           4.4            1.5           0.4   
16    17            5.4           3.9            1.3           0.4   
17    18            5.1           3.5            1.4           0.3   
18    19            5.7           3.8            1.7           0.3   
19    20            5.1           3.8            1.5           0.3   
20    21            5.4           3.4            1.7           0.2   
21    22            5.1           3.7            1.5           0.4   
22    23            4.6           3.6            1.0           0.2   
23    24            5.1           3.3            1.7           0.5   
24    25            4.8           3.4            1.9           0.2   
25    26            5.0           3.0            1.6           0.2   
26    27            5.0           3.4            1.6           0.4   
27    28            5.2           3.5            1.5           0.2   
28    29            5.2           3.4            1.4           0.2   
29    30            4.7           3.2            1.6           0.2   
..   ...            ...           ...            ...           ...   
120  121            6.9           3.2            5.7           2.3   
121  122            5.6           2.8            4.9           2.0   
122  123            7.7           2.8            6.7           2.0   
123  124            6.3           2.7            4.9           1.8   
124  125            6.7           3.3            5.7           2.1   
125  126            7.2           3.2            6.0           1.8   
126  127            6.2           2.8            4.8           1.8   
127  128            6.1           3.0            4.9           1.8   
128  129            6.4           2.8            5.6           2.1   
129  130            7.2           3.0            5.8           1.6   
130  131            7.4           2.8            6.1           1.9   
131  132            7.9           3.8            6.4           2.0   
132  133            6.4           2.8            5.6           2.2   
133  134            6.3           2.8            5.1           1.5   
134  135            6.1           2.6            5.6           1.4   
135  136            7.7           3.0            6.1           2.3   
136  137            6.3           3.4            5.6           2.4   
137  138            6.4           3.1            5.5           1.8   
138  139            6.0           3.0            4.8           1.8   
139  140            6.9           3.1            5.4           2.1   
140  141            6.7           3.1            5.6           2.4   
141  142            6.9           3.1            5.1           2.3   
142  143            5.8           2.7            5.1           1.9   
143  144            6.8           3.2            5.9           2.3   
144  145            6.7           3.3            5.7           2.5   
145  146        

In [24]:
iris.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 6 columns):
Id               150 non-null int64
SepalLengthCm    150 non-null float64
SepalWidthCm     150 non-null float64
PetalLengthCm    150 non-null float64
PetalWidthCm     150 non-null float64
Species          150 non-null object
dtypes: float64(4), int64(1), object(1)
memory usage: 7.1+ KB


In [25]:
iris_2=iris.groupby('PetalLengthCm')['Species']

In [26]:
iris_2.value_counts()

PetalLengthCm  Species        
1.0            Iris-setosa         1
1.1            Iris-setosa         1
1.2            Iris-setosa         2
1.3            Iris-setosa         7
1.4            Iris-setosa        12
1.5            Iris-setosa        14
1.6            Iris-setosa         7
1.7            Iris-setosa         4
1.9            Iris-setosa         2
3.0            Iris-versicolor     1
3.3            Iris-versicolor     2
3.5            Iris-versicolor     2
3.6            Iris-versicolor     1
3.7            Iris-versicolor     1
3.8            Iris-versicolor     1
3.9            Iris-versicolor     3
4.0            Iris-versicolor     5
4.1            Iris-versicolor     3
4.2            Iris-versicolor     4
4.3            Iris-versicolor     2
4.4            Iris-versicolor     4
4.5            Iris-versicolor     7
               Iris-virginica      1
4.6            Iris-versicolor     3
4.7            Iris-versicolor     5
4.8            Iris-versicolor     2
       

In [27]:
pd.crosstab(iris['PetalLengthCm'],iris['Species'])

Species        Iris-setosa  Iris-versicolor  Iris-virginica
PetalLengthCm                                              
1.0                      1                0               0
1.1                      1                0               0
1.2                      2                0               0
1.3                      7                0               0
1.4                     12                0               0
1.5                     14                0               0
1.6                      7                0               0
1.7                      4                0               0
1.9                      2                0               0
3.0                      0                1               0
3.3                      0                2               0
3.5                      0                2               0
3.6                      0                1               0
3.7                      0                1               0
3.8                      0                1               0
3.9                      0                3               0
4.0                      0                5               0
4.1                      0                3               0
4.2                      0                4               0
4.3                      0                2               0
4.4                      0                4               0
4.5                      0                7               1
4.6                      0                3               0
4.7                      0                5               0
4.8                      0                2               2
4.9                      0                2               3
5.0                      0                1               3
5.1                      0                1               7
5.2                      0                0               2
5.3                      0                0               2
5.4                      0                0               2
5.5                      0                0               3
5.6                      0                0               6
5.7                      0                0               3
5.8                      0                0               3
5.9                      0                0               2
6.0                      0                0               2
6.1                      0                0               3
6.3                      0                0               1
6.4                      0                0               1
6.6                      0                0               1
6.7                      0                0               2
6.9                      0                0               1

# As we can  clearly  see  that all Iris Setosa has PetalLength less than 2.0 and so we can conclude that  the first split should take place at PL>1.9 .    